includes/embeds/class-amp-youtube-embed-handler.php

```diff
@@ -77,6 +80,90 @@ public function unregister_embed() {
 		remove_filter( 'embed_oembed_html', [ $this, 'filter_embed_oembed_html' ], 10 );
 	}
 
+	/**
+	 * Sanitize HTML that are not added via Gutenberg.
+	 *
+	 * @param Document $dom  Document.
+	 *
+	 * @return void
+	 */
+	public function sanitize_raw_embeds( Document $dom ) {
+
+		$applicable_domains = [ 'youtu.be', 'youtube.com', 'youtube-nocookie.com' ];
+
+		$query_segments = array_map(
+			static function ( $domain ) {
+
+				return sprintf(
+					'starts-with( @src, "https://www.%1$s/" ) or starts-with( @src, "https://%1$s/" ) or starts-with( @src, "http://www.%1$s/" ) or starts-with( @src, "http://%1$s/" )',
+					$domain
+				);
+			},
+			$applicable_domains
+		);
+
+		$query = implode( ' or ', $query_segments );
+
+		$nodes = $dom->xpath->query( sprintf( '//iframe[ %s ]', $query ) );
+
+		/** @var Element $node */
+		foreach ( $nodes as $node ) {
+
+			$html             = $dom->saveHTML( $node );
+			$url              = $node->getAttribute( 'src' );
+			$amp_youtube_node = $this->process_embed( $dom, $html, $url );
+
+			if ( ! empty( $amp_youtube_node ) ) {
+				$node->parentNode->replaceChild( $amp_youtube_node, $node );
+			}
+		}
+
+	}
+
+	/**
+	 * To AMP youtube component from DOM Document.
+	 *
+	 * @param Document $dom  Document DOM.
+	 * @param string   $html HTML markup of youtube iframe.
+	 * @param string   $url  Youtube URL.
+	 *
+	 * @return DOMElement|false DOMElement on success, Otherwise false.
+	 */
+	public function process_embed( Document $dom, $html, $url ) {
+
+		$id = $this->get_video_id_from_url( $url );
```

Comment:
I just realized the AMP component can also [accept a YouTube channel id](Youtube channel id), so we need to provide support for that as well. So if there is no video ID, then check for a live channel ID.

From the docs:

> For example, in this URL: `https://www.youtube.com/embed/live_stream?channel=UCB8Kb4pxYzsDsHxzBfnid4Q`, `UCB8Kb4pxYzsDsHxzBfnid4Q` is the channel id. You can provide a `data-live-channelid` instead of a `data-videoid` attribute to embed a stable url for a live stream instead of a video.

Reply:
@pierlon I have added support for live streaming channels for iframe added as custom HTML markup.
However, WordPress by default does not provide support for embedding YouTube URLs (It will just render as text)

Reply:
If WordPress never supported it I think it's safe to say we don't have to either.

> @pierlon I have added support for live streaming channels for iframe added as custom HTML markup.

I tried this but it parsed an incorrect video ID. Eg:

```html
<iframe src="https://www.youtube.com/embed/live_stream?channel=12345" allowfullscreen="" width="560" height="315"></iframe>
```

Gives:

```html
<amp-youtube data-videoid="live_stream" layout="responsive" width="560" height="315" data-param-channel="12345" ...</amp-youtube>
```

Instead, the `data-videoid` attribute should not have been set, nor the `data-param-channel` attribute. Perhaps we could add a condition to say if the parsed video ID is `live_stream` then it's invalid and bail.

Reply:
Have addressed this issue.